demos/common/monitors/cpu_monitor.cpp

```diff
@@ -0,0 +1,274 @@
+// Copyright (C) 2019 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "cpu_monitor.h"
+#ifdef WIN32
+#include <tchar.h>
+
+namespace {
+std::size_t getNCores() {
+    SYSTEM_INFO sysinfo;
+    GetSystemInfo(&sysinfo);
+    return sysinfo.dwNumberOfProcessors;
+}
+}
+
+CpuMonitor::CpuMonitor() : nCores{getNCores()}, lastEnabled{false}, historyEnabled{false}, samplesNumber{0} {};
+
+void CpuMonitor::openQuery() {
+    PDH_STATUS status;
+
+    status = PdhOpenQuery(NULL, NULL, &query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhOpenQuery() failed");
+    }
+
+    for (std::size_t i = 0; i < nCores; ++i)
+    {
+        TCHAR szCounterName[MAX_PATH];
+        _stprintf_s(szCounterName, sizeof(szCounterName), TEXT("\\Processor(%u)\\%% Processor Time"), i);
+
+        PDH_HCOUNTER counter;
+
+        status = PdhAddCounter(query, szCounterName, 0, &counter);
+        if (ERROR_SUCCESS != status)
+        {
+            throw std::runtime_error("PdhAddCounter() failed");
+        }
+
+        status = PdhSetCounterScaleFactor(counter, -2); // scale counter to [0, 1]
+        if (ERROR_SUCCESS != status)
+        {
+            throw std::runtime_error("PdhSetCounterScaleFactor() failed");
+        }
+
+        coreTimeCounters.push_back(counter);
+    }
+    status = PdhCollectQueryData(query);
+    if (ERROR_SUCCESS != status)
+    {
+        throw std::runtime_error("PdhCollectQueryData() failed");
+    }
+}
+
+void CpuMonitor::closeQuery() {
+    PDH_STATUS status = PdhCloseQuery(query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhCloseQuery() failed");
+    }
+}
+
+void CpuMonitor::enableHistory(std::size_t historySize) {
+    if (0 == historySize) {
+        disableHistory();
+    }
+    else {
+        this->historySize = historySize;
+        historyEnabled = true;
```

Comment:
Isn't `historyEnabled` basically equivalent to `historySize != 0`? Why keep it as a separate variable?

Reply:
`historySize == 1` can mean that history is enabled with length 1, but it also can mean that only last value is enabled.

Reply:
What's the difference?

Reply:
Having `historySize == 1` (which is != 0) still can mean that history is disabled.

Reply:
Yeah, but what is the actual difference in behavior between history being enabled and disabled when `historySize == 1`?

Reply:
There is no reason to draw empty graph if `historySize == 1`, so `historySize < 2` can be treated as disabled history. Updated